enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,821 @@
+to the ACL rule(s) for the EgressFirewall rule(s) in which the DNS name is used. This will ensure that the latest IP addresses are always updated in
+the `AddressSets`.
+
+#### Cluster Network Operator
+
+To add the support of `DNSNameResolver` CRD in OVN-K, add the flag `--enable-dns-name-resolver` to the corresponding OVN-K pods.
+
+
+### Risks and Mitigations
+
+* The `DNSNameResolver` CR will be created by OVN-K Cluster manager whenever a new DNS name is used in a EgressFirewall rule. The CR will be deleted
+when the corresponding DNS name is not used in any of the EgressFirewall rules. The `DNSNameResolver` CR should not be modified (created or deleted or
+updated) by a user. Doing so may lead to undesired behavior of EgressFirewall.
+* During the upgrade process if the OVN-K cluster manager and master pods are upgraded first and then the Cluster DNS operator and the CoreDNS pods, the
+EgressFirewall functionality may break. The details can be found in [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy) section
+
+### Drawbacks
+
+* Whenever there's a change in the IP addresses or the next lookup time (TTL + last lookup time) for a DNS name, the additional step of updating the
+related `DNSNameResolver` CRs will be executed. This will add some delay to the DNS lookup process. However, this will only
+happen whenever there's a change in the DNS information.
+* For wildcard DNS names, it is not possible to pre-emptively find out the matching regular DNS names which have DNS records in the upstream nameservers.
+For these regular DNS names the first lookup may originate from a workload pod, in which case the pod may
+see an initial delay until the underlying ACL rules are updated with the corresponding IP addresses. However, this will happen only for the first DNS lookup
+of a particular DNS name by any pod. Subsequently, it will be handled by the `DNSNameResolver` controller to keep the IP addresses updated.
+
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
+
+* This enhancement will be tested through e2e tests by adding EgressFirewall rules containing regular DNS names
+and wildcard DNS names. The tests will be added to the `openshift/origin` repository.
+* Testing the feature where IP addresses are changed may be a little bit tricky as this will probably include
+creation of DNS records and then changing the IP addresses of the DNS records through the e2e tests.
+
+### Graduation Criteria
+
+This feature will initially be released as Tech Preview only.
+
+#### Dev Preview -> Tech Preview
+
+N.A. This feature will go directly to Tech Preview.
+
+#### Tech Preview -> GA (Future work)
+
+* Incorporate the feedback received on the Tech Preview version.
+* OpenShift documentation is needed to be updated.
+* UTs and e2e tests need to cover all the edge case scenarios.
+
+#### Removing a deprecated feature
+
+
+### Upgrade / Downgrade Strategy
```

Comment:
Adding/removing the plugin from CoreDNS will be interesting - need tests to make sure it doesn't greatly impact the upgrade/downgrade times.

Reply:
Will cover the scenario in e2e tests.